src/ZODB/tests/BasicStorage.py

```diff
@@ -385,3 +388,237 @@ def lastInvalidations():
+        def init():
+            db = dbopen()
+
+            transaction.begin()
+            zconn = db.open()
+
+            root = zconn.root()
+            root['obj1'] = MinPO(0)
+            root['obj2'] = MinPO(0)
+
+            transaction.commit()
+            zconn.close()
+
+            db.close()
+
+
+        # T is a worker that accesses obj1/obj2 in a loop and verifies
+        # `obj1.value == obj2.value` invariant.
+        #
+        # access to obj1 is organized to always trigger loading from zstor.
+        # access to obj2 goes through zconn cache and so verifies whether the cache is not stale.
+        #
+        # Once in a while T tries to modify obj{1,2}.value maintaining the invariant as
+        # test source of changes for other workers.
+        failed = threading.Event()
+        failure = [] # [tx] is failure from T(tx)
+        def T(tx, N):
+            db = dbopen()
+
+            def t_():
+                transaction.begin()
+                zconn = db.open()
+
+                root = zconn.root()
+                obj1 = root['obj1']
+                obj2 = root['obj2']
+
+                # obj1 - reload it from zstor
+                # obj2 - get it from zconn cache
+                obj1._p_invalidate()
+
+                # both objects must have the same values
+                i1 = obj1.value
+                i2 = obj2.value
+                if i1 != i2:
+                    #print('FAIL')
+                    failure[tx] = "T%s: obj1.value (%d)  !=  obj2.value (%d)" % (tx, i1, i2)
+                    failed.set()
+
+                # change objects once in a while
+                if randint(0,4) == 0:
+                    #print("T%s: modify" % tx)
+                    obj1.value += 1
+                    obj2.value += 1
+
+                try:
+                    transaction.commit()
+                except POSException.ConflictError:
+                    #print('conflict -> ignore')
+                    transaction.abort()
+
+                zconn.close()
+
+            try:
+                for i in range(N):
+                    #print('T%s.%d' % (tx, i))
+                    t_()
+                    if failed.is_set():
+                        break
+            except:
+                failed.set()
+                raise
+            finally:
+                db.close()
+
+
+        # run 8 T workers concurrently. As of 20210416, due to race conditions
+        # in ZEO, it triggers the bug where T sees stale obj2 with obj1.value != obj2.value
+        #
+        # The probability to reproduce the bug is significantly reduced with
+        # decreasing n(workers): almost never with nwork=2 and sometimes with nwork=4.
+        nwork = 8
+        init()
+
+        N = 100
+        tg = []
+        failure = [None]*nwork
+        for x in range(nwork):
+            t = threading.Thread(name='T%d' % x, target=T, args=(x, N))
+            t.start()
+            tg.append(t)
```

Comment:
what about
```py
tg = [threading.Thread(name='T%d' % x, target=T, args=(x, N))
      for x in range(nwork)]
for t in tg:
    t.start()

Reply:
it is a bit less clear to me, so I prefer to leave my version of this code.

Reply:
Besides style (here, I won't insist), I wonder it makes a difference to minimize the difference of start time between threads, and if it would have an impact to reproduce failure. Or maybe it's completely negligible.

Reply:
I've tried this change and run the tests on unfixed ZEO. The failure is reproducible with both versions. The iteration number at which the failue is reproduced is scattered in between 1-5-6-30-... in both versions. So I believe this change is completely orthogonal to test semantic.